Turning text-spacing properties into font properties

vitorroriz opened the pull request:
#### 61c1b82983f07f644c9e08fc37906f19037fafff
<pre>
Turning text-spacing properties into font properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=252660">https://bugs.webkit.org/show_bug.cgi?id=252660</a>
rdar://105722440

Reviewed by Antti Koivisto.

We need turn text-spacing properties into font properties such that
they are accessible during FontCascade width calculations.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::textSpacingTrim const):
(WebCore::FontCascade::textAutospace const):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialTextSpacingTrim):
(WebCore::FontCascadeDescription::initialTextAutospace):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::textSpacingTrim const):
(WebCore::FontDescription::textAutospace const):
(WebCore::FontDescription::setTextSpacingTrim):
(WebCore::FontDescription::setTextAutospace):
(WebCore::FontDescription::operator== const):
(WebCore::FontDescription::encode const):
(WebCore::FontDescription::decode):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::textSpacingTrim const):
(WebCore::RenderStyle::textAutospace const):
(WebCore::RenderStyle::setTextSpacingTrim):
(WebCore::RenderStyle::setTextAutospace):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialTextOrientation):
(WebCore::RenderStyle::setTextSpacingTrim): Deleted.
(WebCore::RenderStyle::setTextAutospace): Deleted.
(WebCore::RenderStyle::initialTextSpacingTrim): Deleted.
(WebCore::RenderStyle::initialTextAutospace): Deleted.
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: <a href="https://commits.webkit.org/267224@main">https://commits.webkit.org/267224@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0ed05f6224e063cca4a9363fe15b9056442631ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16010 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16328 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16720 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17774 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15031 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16197 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16432 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17472 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13659 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18536 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13918 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14475 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/21337 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14907 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14641 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17884 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15236 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12918 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14473 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3821 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18842 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15057 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2104,6 +2104,26 @@ void RenderStyle::setLetterSpacing(float letterSpacing)
+void RenderStyle::setTextSpacingTrim(TextSpacingTrim value)
+{
+    auto selector = fontCascade().fontSelector();
+    auto description = fontDescription();
```
Can we mutate the `fontDescription()` directly with a `auto&` (instead of copy + move) ?

vitorroriz replied:
setFontDescription() accepts only a rvalue reference for the description object as argument

mdubet replied:
But we don't need to call `setFontDescription()` if we take a reference and mutate it ?

vitorroriz replied:
I believe the description is used to construct a new FontCascade at setFontDescription(), which mutates the FontCascade object stored by Style. FontCascade has only constructors that get a rvalue FontDescription. I think that's why all other functions are following the pattern.

mdubet replied:
Thanks for the explanation. It seems we could use a macro for those font members (like the `SET_VAR` macro) (not in this PR obviously).

anttijk commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -177,6 +182,8 @@ class FontDescription {
 
     FontSelectionRequest m_fontSelectionRequest;
     Markable<float, FloatMarkableTraits> m_sizeAdjust; // Size adjust for font-size-adjust
+    TextSpacingTrim m_textSpacingTrim;
+    TextAutospace m_textAutospace;
```
These should go to the bitfield to avoid bloating the type.

vitorroriz replied:
The longhands should support future '||' syntax with too many combinations for representing it as enum in a readable way:

`text-autospace: normal | auto | no-autospace | [ [ ideograph-alpha || ideograph-numeric || punctuation ] || [ insert | replace ] ]"`

anttijk replied:
Not sure I see the relation to my comment.

anttijk replied:
I mean you can put an OptionSet to a bitfield too.

vitorroriz committed:
Turning text-spacing properties into font properties
https://bugs.webkit.org/show_bug.cgi?id=252660
rdar://105722440

Reviewed by Antti Koivisto.

We need turn text-spacing properties into font properties such that
they are accessible during FontCascade width calculations.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::textSpacingTrim const):
(WebCore::FontCascade::textAutospace const):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
(WebCore::FontCascadeDescription::initialTextSpacingTrim):
(WebCore::FontCascadeDescription::initialTextAutospace):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::textSpacingTrim const):
(WebCore::FontDescription::textAutospace const):
(WebCore::FontDescription::setTextSpacingTrim):
(WebCore::FontDescription::setTextAutospace):
(WebCore::FontDescription::operator== const):
(WebCore::FontDescription::encode const):
(WebCore::FontDescription::decode):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::textSpacingTrim const):
(WebCore::RenderStyle::textAutospace const):
(WebCore::RenderStyle::setTextSpacingTrim):
(WebCore::RenderStyle::setTextAutospace):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::initialTextOrientation):
(WebCore::RenderStyle::setTextSpacingTrim): Deleted.
(WebCore::RenderStyle::setTextAutospace): Deleted.
(WebCore::RenderStyle::initialTextSpacingTrim): Deleted.
(WebCore::RenderStyle::initialTextAutospace): Deleted.
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:

Canonical link: https://commits.webkit.org/267224@main
